slasher/db/kv/spanner.go

```diff
@@ -0,0 +1,306 @@
+package kv
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/boltdb/bolt"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/slasher/detection/attestations/types"
+	log "github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+)
+
+var highestEpoch uint64
+
+func cacheTypeMismatchError(value interface{}) error {
+	return fmt.Errorf("cache contains a value of type: %v "+
+		"while expected to contain only values of type : map[uint64]types.Span", reflect.TypeOf(value))
+}
+
+func saveToDB(db *Store) func(uint64, uint64, interface{}, int64) {
+	// Returning the function here so we can access the DB properly from the OnEvict.
+	return func(epoch uint64, _ uint64, value interface{}, cost int64) {
+		log.Tracef("evicting span map for epoch: %d", epoch)
+		err := db.update(func(tx *bolt.Tx) error {
+			bucket := tx.Bucket(validatorsMinMaxSpanBucket)
+			epochBucket, err := bucket.CreateBucketIfNotExists(bytesutil.Bytes8(epoch))
+			if err != nil {
+				return err
+			}
+			spanMap, ok := value.(map[uint64]types.Span)
+			if !ok {
+				return cacheTypeMismatchError(value)
+			}
+			for k, v := range spanMap {
+				err = epochBucket.Put(bytesutil.Bytes8(k), marshalSpan(v))
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			log.Errorf("failed to save span map to db on cache eviction: %v", err)
+		}
+	}
+}
+
+func unmarshalSpan(ctx context.Context, enc []byte) (types.Span, error) {
+	ctx, span := trace.StartSpan(ctx, "SlasherDB.unmarshalSpan")
+	defer span.End()
+	r := types.Span{}
+	if len(enc) != 7 {
+		return r, errors.New("wrong data length for min max span")
+	}
+	r.MinSpan = bytesutil.FromBytes2(enc[:2])
+	r.MaxSpan = bytesutil.FromBytes2(enc[2:4])
+	sigB := [2]byte{}
+	copy(sigB[:], enc[4:6])
+	r.SigBytes = sigB
+	r.HasAttested = bytesutil.ToBool(enc[6])
+	return r, nil
+}
+
+// convert the span struct into a flat array
+func marshalSpan(span types.Span) []byte {
+	return append(append(append(
+		bytesutil.Bytes2(uint64(span.MinSpan)),
+		bytesutil.Bytes2(uint64(span.MaxSpan))...),
+		span.SigBytes[:]...),
+		bytesutil.FromBool(span.HasAttested))
```

Comment:
Can you newline the last parentheses here and put a trailing comma?

Reply:
added.
thanks